### PR TITLE
[https://nvbugs/5322131][feat] Multi-LoRA serving with CUDA Graph

### PR DESCRIPTION
**Pull request by JyChang012**
# NvBug
https://nvbugspro.nvidia.com/bug/5322131
https://nvbugspro.nvidia.com/bug/5441746

# Benchmark Perf 

merge base `43c46a09`

| | Avg ITL (ms) | Throughput (tok/s) | Avg TTFT (ms) |
|--------|--------|--------|----------|
| CUDA Graph + multi LoRA (after changes) | 26 | 568 | 227 | 
| no CUDA Graph + multi LoRA (before changes) | 119 | 127 | 435 |
| CUDA Graph + no LoRA (before changes) | 14 | 1068 | 137 |

Llama 3.3 70B, TP8; p5.48xlarge 8xH100
ISL: 1600; OSL: 600; Concurrency: 16; All requests query the same LoRA

Still need to remove some code for logging / testing 

# Potential Future Optimizations not included in this PR 
- Update prefill + decode fused batch to the new LoRA path, which might reduce bubbles in all-reduce
- Update the sm80 (split-K) group GEMMs currently used

## Summary by CodeRabbit

- New Features
  - Added CUDA Graph mode for LoRA with multi-adapter batching and slot management.
  - Introduced fused parameter preparation and row reordering to reduce kernel launches.
  - Exposed a device-side cache check for tasks in Python.
  - Enabled optional NVTX profiling wrappers for easier performance tracing.

- Performance
  - Implemented CUDA Graph–compatible grouped and split-K GEMM paths for faster LoRA execution.
  - Reduced per-step overhead via persistent buffers and slot reuse.

- Tests
  - Expanded test coverage to run LoRA scenarios with and without CUDA Graph, including edge cases.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/modules/attention.py [hunk 1]
```diff
@@ -272,9 +272,10 @@ def __init__(
         ], [self.q_size, self.kv_size, self.kv_size])
         self.fused_qkv_lora = LoraLayer([LoraModuleType.ATTENTION_QKV],
                                         [self.q_size + 2 * self.kv_size])
-
+        '''
         self.o_lora = LoraLayer([LoraModuleType.ATTENTION_DENSE],
                                 [self.hidden_size])
+        '''
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply tensorrt_llm/_torch/modules/attention.py]: Isn't it not used? It was set before and passed to `self.o_proj`, and reset here and not used anymore.

File: cpp/tensorrt_llm/thop/loraOp.cpp [hunk 2]
```diff
@@ -172,6 +175,197 @@ std::vector<th::Tensor> lora_grouped_gemm(th::Tensor const& input, th::Tensor co
...
+    if (problem_count > 0)
+    {
+        TLLM_LOG_TRACE("Start Grouped GEMM for LoRA in.");
+
+        tk::cuda_graph_splitk_grouped_gemm(problem_sizes_1_ptr, problem_count, a_ptrs_gpu, b_ptrs_gpu,
+            d_ptrs_gpu,                                     // ptrC (no bias)
+            d_ptrs_gpu, lda_gpu, ldb_gpu, ldd_gpu, ldd_gpu, // Precomputed leading dimensions
+            true,                                           // isLoraIn
+            loraRuntimeDataType,
+            16,                                             // splitKSlices
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply cpp/tensorrt_llm/thop/loraOp.cpp]: It was hard-coded [here](https://github.com/NVIDIA/TensorRT-LLM/blob/65ec01b257fdcf2fc460447e62c3e296ae926952/cpp/tensorrt_llm/kernels/lora/lora.h#L64) before

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    }
+    else if (threads_per_block == 288)
+    {
+        launchKernelWithModuleCount<288>(in_sizes, out_sizes, a_ptrs, d_ptrs, d_prime_ptrs, lda, ldd, ldb_prime,
+            ldd_prime, splitk_offsets, reordered_input, max_lora_count, max_lora_rank, sum_output_hidden_size,
+            input_hidden_size, dtype_element_size, batch_size, a_base, d_base, d_prime_base, slot_counts, slot_ranks,
+            slot_offsets, module_out_sizes, module_out_prefix, b_ptrs, b_prime_ptrs, input, sorted_ids, module_count,
+            stream);
+    }
+    else if (threads_per_block == 320)
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: I make threads per block multiples of 32 which is the warp size and its min to be 256. And this number should be larger than max_lora_count * module_count. Max of the latter currently is 3, right? So I kind of implicitly enforce max lora count here.

File: cpp/tensorrt_llm/batch_manager/peftCacheManager.cpp [hunk 4]
```diff
@@ -462,10 +462,8 @@ PeftCacheManager::PeftTable PeftCacheManager::ensureBatch(
     {
         auto&& f = ensureFutures.at(taskId);
         auto const values = f.get();
-        for (auto const& reqId : reqIds)
-        {
-            peftTable.try_emplace(reqId, values);
-        }
+        // Map task_id to layer-module-configs instead of request_id to layer-module-configs
+        peftTable.try_emplace(taskId, values);
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply cpp/tensorrt_llm/batch_manager/peftCacheManager.cpp]: You are right. Let me fix it.

File: cpp/tensorrt_llm/thop/loraOp.cpp [hunk 5]
```diff
@@ -172,6 +175,197 @@ std::vector<th::Tensor> lora_grouped_gemm(th::Tensor const& input, th::Tensor co
...
+            minKnInt,                                       // minKN
+            host_max_in_sizes_ptr, splitk_offsets_gpu, stream);
+        sync_check_cuda_error(stream);
+    }
+
+    // Step 7: Call CUDA Graph compatible grouped GEMM for lora_out
+    if (problem_count > 0)
+    {
```
Comment (amitz-nv) [inline cpp/tensorrt_llm/thop/loraOp.cpp]:
```suggestion

        // Step 7: Call CUDA Graph compatible grouped GEMM for lora_out
```
If accepted, I would also move the `// Step 6:` comment to be inside the `if`

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    D_PRIME_PTRS_INDEX,
+    SPLITK_OFFSETS_INDEX,
+    PARAM_COUNT,
+};
+
+int constexpr VECTOR_LOAD_WIDTH = 16;
+} // namespace
+
+/**
+ * Fused kernel for LoRA group GEMM parameter filling, row gather and zero fillings.
```
[the remark this answers is not in the document]
Comment (amitz-nv) [reply cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: I think that adding to this documentation what reordering is done and why can be helpful

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    case D_PTRS_INDEX:
+    case D_PRIME_PTRS_INDEX: break;
+    default:
+        if (linearIdx == 0)
+        {
+            row_count_to_gather = static_cast<int32_t>(slot_offsets[max_lora_count]);
+        }
+    }
+
+    int constexpr ITEM_PER_THREAD = VECTOR_LOAD_WIDTH / sizeof(uint8_t);
```
Comment (amitz-nv) [inline cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: What's the reason behind the division by `sizeof(uint8_t)` here and in other calculations in this function?

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (amitz-nv) [reply cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: I see, then what's the 512 for?

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu (hunk 7 above)
Comment (JyChang012) [reply cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: Mainly to convert all sizes to bytes and use cub's 128bit vector load / store primitives for different types

File: cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h [hunk 8]
```diff
@@ -58,6 +58,9 @@ class BasePeftCacheManager
     using LlmRequestPtr = std::shared_ptr<LlmRequest>;
     using RequestVector = std::vector<LlmRequestPtr>;
     using PeftTable = std::map<uint64_t, std::vector<runtime::LoraCache::TaskLayerModuleConfig>>;
+    using TaskPeftTable = std::map<uint64_t, std::vector<runtime::LoraCache::TaskLayerModuleConfig>>;
+    using TaskIdToReqIds = std::map<uint64_t, std::vector<uint64_t>>;
```
Comment (Funatiq) [inline cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h]: Can we use `unordered_map`s here?

File: cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu (hunk 7 above)
Comment (Funatiq) [reply cpp/tensorrt_llm/kernels/lora/loraGroupGEMMParamFillRowReorderFusion.cu]: Isn't `sizeof(uint8_t)` always equal to `1`?

**Review by Funatiq — COMMENTED** [review]
Does the change also affect the runtime w/o CUDA graphs? (`no CUDA Graph + multi LoRA (after changes)`)

File: tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __init__(self, max_lora_size: int, device: str = "cuda"):
+        """
+        Initialize the AdapterSlotManager.
+
+        Args:
+            max_lora_size: Maximum number of LoRA adapters that can be active simultaneously
+            device: Device to allocate tensors on
+        """
+        self.max_lora_size = max_lora_size
```
Comment (amitz-nv) [inline tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py]: I think renaming `max_lora_size` so it would better reflect that it's the max num of lora adapters would help make this class easier to comprehend

File: tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            Dict mapping request_id to slot_id, with slot_id=max_lora_size for base model requests
+        """
+        # remove task evicted in PeftCacheManager in C++
+        self.remove_evicted_slots_in_cpp(peft_cache_manager)
+
+        # check if total number of unique tasks in the requests is not larger than max_lora_size
+        tasks = [request.lora_task_id for request in requests]
+        unique_tasks = {t for t in tasks if t is not None}
+        assert len(unique_tasks) <= self.max_lora_size, (
+            "Currently do not support batch with more LoRA task ID than loRA slot size!"
```
Comment (amitz-nv) [inline tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py]:
```suggestion
            "Batch with more unique LoRA adapters ({len(unique_tasks)}) than max_lora_size={self.max_lora_size} is not supported"
```
or something similar, as I don't think the slot size is relevant here, or maybe I misunderstood something?

File: tests/unittest/llmapi/test_llm_pytorch.py [hunk 11]
```diff
@@ -261,19 +263,74 @@ def test_embedding_bias_with_torch_sampler_strategies():
...
-    llm = LLM(
-        model=f"{llm_models_root()}/llama-models/llama-7b-hf",
-        lora_config=lora_config,
-        # Disable CUDA graph
-        # TODO: remove this once we have a proper fix for CUDA graph in LoRA
-        cuda_graph_config=None,
-        **llm_kwargs)
+    llm = LLM(model=f"{llm_models_root()}/llama-models/llama-7b-hf",
+              lora_config=lora_config,
+              **llm_kwargs)
```
Comment (amitz-nv) [inline tests/unittest/llmapi/test_llm_pytorch.py]: Note that this change should affect all calls to `llama_7b_lora_from_dir_test_harness`, which is also called from `tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py`.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 12]
```diff
@@ -1224,7 +1250,9 @@ def _prepare_tp_inputs(
             attn_metadata: AttentionMetadata,
             spec_metadata: Optional[SpecMetadata] = None,
             new_tensors_device: Optional[SampleStateTensors] = None,
-            cache_indirection_buffer: Optional[torch.Tensor] = None):
+            cache_indirection_buffer: Optional[torch.Tensor] = None,
+            maybe_graph: bool = False,
```
Comment (amitz-nv) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: I'm not entirely sure what this means, does this control whether CUDA graphs should be used or not? 
Can you change its name to be more explicit? (here and anywhere else this name is used)

File: cpp/tensorrt_llm/thop/loraOp.cpp [hunk 13]
```diff
@@ -172,6 +175,173 @@ std::vector<th::Tensor> lora_grouped_gemm(th::Tensor const& input, th::Tensor co
...
+            loraRuntimeDataType,
+            16,                                             // splitKSlices
+            minKnInt,                                       // minKN
+            host_max_in_sizes_ptr, splitk_offsets_gpu, stream);
+        sync_check_cuda_error(stream);
+    }
+
+    // Call CUDA Graph compatible grouped GEMM for lora_out
+    if (problem_count > 0)
+    {
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply cpp/tensorrt_llm/thop/loraOp.cpp]: Done

File: tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py (hunk 9 above)
Comment (JyChang012) [reply tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py]: done

File: tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py (hunk 10 above)
Comment (JyChang012) [reply tensorrt_llm/_torch/peft/lora/adapter_slot_manager.py]: done

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 14]
```diff
@@ -1435,17 +1439,10 @@ def prepare_resources(self, scheduled_batch: ScheduledRequests):
...
 
-        for req in context_batch:
-            req.py_lora_task_layer_module_configs = py_lora_task_layer_module_configs[
-                req.
-                py_request_id] if req.py_request_id in py_lora_task_layer_module_configs else None
-        for req in generation_batch:
-            req.py_lora_task_layer_module_configs = py_lora_task_layer_module_configs[
-                req.
-                py_request_id] if req.py_request_id in py_lora_task_layer_module_configs else None
+        # torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py]: done

File: cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h (hunk 8 above)
Comment (Funatiq) [reply cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h]: If we need the ordering than we have to use `std::map`. If not then we should prefer `std:: unordered_map` because of O(1) insert and access.

File: tests/unittest/llmapi/test_llm_pytorch.py (hunk 11 above)
Comment (JyChang012) [reply tests/unittest/llmapi/test_llm_pytorch.py]: set default to none

File: tests/unittest/llmapi/test_llm_pytorch.py [hunk 15]
```diff
@@ -295,22 +352,21 @@ def llama_7b_lora_from_dir_test_harness(**llm_kwargs) -> None:
...
-def test_llama_7b_lora():
-    llama_7b_lora_from_dir_test_harness()
+@test_lora_with_and_without_cuda_graph
+def test_llama_7b_lora(cuda_graph_config):
+    llama_7b_lora_from_dir_test_harness(cuda_graph_config=cuda_graph_config)
 
 
 @skip_gpu_memory_less_than_40gb
-def test_llama_7b_lora_default_modules() -> None:
+@test_lora_with_and_without_cuda_graph
```
[the remark this answers is not in the document]
Comment (JyChang012) [reply tests/unittest/llmapi/test_llm_pytorch.py]: What is the incorrect output? I cannot reproduce this error

File: tests/unittest/llmapi/test_llm_pytorch.py
[the remark this answers is not in the document]
Comment (JyChang012) [reply tests/unittest/llmapi/test_llm_pytorch.py]: Added multi GPU tests. After a quick look I do not think we have much to omit in single GPU test as many of them test the eviction / reloading logic which is heavily modified in CUDA Graph path and had issues before.

**Review by amitz-nv — CHANGES_REQUESTED** [review]
Impressive work!

Please:
1. Fix the very minor comments remaining.
2. Fix merge conflicts.
3. Get a passing CI run - make sure it runs the CPP tests as well (I'm not sure whether it happens by default).

**Review by amitz-nv — APPROVED** [review]
Impressive work! Overall the code looks good to me

Note that since this PR is very large, and modifies all software "levels" (CUDA kernels, CPP, python) I couldn't review everything as thoroughly as I would like and as this feature deserves, but I believe that at this stage if the CI passes successfully then it should be OK.

Also, I think it can be nice as an additional documentation if the PR description would contain a summary of the high level changes introduced in this PR (for example: The new CUDA kernels and their purpose, some data structure changes, etc.)

**Review by eopXD — COMMENTED** [review]
> @NVIDIA/trt-llm-kv-cache-manager-devs Could you please approve this PR? We mainly added new member functions in PeftCacheManagerBinding in cpp/tensorrt_llm/pybind/batch_manager/kvCacheManager.cpp and nanobind.

- What are the functions for and its impact? What is the behavioral change of the new `PeftCacheManager::ensureBatch`?
- New type-defs are introduced, there are both functional fix and structural change (map -> unordered_map) in this MR. May you give an overview of them? Comments in the codebase will also be appreciated.

File: cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h [hunk 16]
```diff
@@ -57,7 +57,10 @@ class BasePeftCacheManager
 public:
     using LlmRequestPtr = std::shared_ptr<LlmRequest>;
     using RequestVector = std::vector<LlmRequestPtr>;
-    using PeftTable = std::map<uint64_t, std::vector<runtime::LoraCache::TaskLayerModuleConfig>>;
+    using PeftTable = std::unordered_map<uint64_t, std::vector<runtime::LoraCache::TaskLayerModuleConfig>>;
```
Comment (JyChang012) [inline cpp/include/tensorrt_llm/batch_manager/peftCacheManager.h]: @eopXD The change to unordered map is suggested by @Funatiq . As we find that ordering is not needed anywhere and unordered map is algorithmically faster.

Adding TaskPeftTable is to indicate the newly added `ensureBatchMapTaskId` return a taskId to layer configs map

Adding `TaskIdToReqIds` is to shorten type declaration

The discussion threads can be found in the `Conversions` dropdown in the `files changed` page, under resolved conversation. They are on longer visible outside after several commit squash

**Review by yuxianq — APPROVED** [review]
The attention part LGTM, approved.

File: tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py [hunk 17]
```diff
@@ -48,7 +49,8 @@ def test_llama_7b_lora_tp2():
 
 @pytest.mark.gpu2
 @pytest.mark.skip(reason="https://nvbugs/5682551")
```
Comment (eopXD) [inline tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py]: Should we remove the skip?

File: tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py [hunk 18]
```diff
@@ -48,7 +49,8 @@ def test_llama_7b_lora_tp2():
...
-def test_llama_7b_multi_lora_tp2():
+@test_lora_with_and_without_cuda_graph
+def test_llama_7b_multi_lora_tp2(cuda_graph_config):
```
Comment (eopXD) [inline tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py]: Test TP4 please.

**Review by eopXD — APPROVED** [review]
Looks good to me, only minor comments on test coverage under `tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py`.

File: tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py (hunk 18 above)
Comment (JyChang012) [reply tests/unittest/llmapi/test_llm_multi_gpu_pytorch.py]: changed to tp4
